examples/seismic/skew_self_adjoint/test_wavesolver_iso.py

```diff
@@ -0,0 +1,542 @@
+            dr . J(dm) = J^T(dr) . dm
+        """
+        omega = 2.0 * np.pi * fpeak
+        b, v, time_axis, src_coords, rec_coords = \
+            defaultSetupIso(npad, shape, dtype, tmax=tmax)
+        nt = time_axis.num
+
+        solver = SSA_ISO_AcousticWaveSolver(npad, qmin, qmax, omega, b, v,
+                                            src_coords, rec_coords, time_axis,
+                                            space_order=so)
+
+        src0 = RickerSource(name='src0', grid=v.grid, f0=fpeak, npoint=1,
+                            time_range=time_axis)
+        src0.coordinates.data[:] = src_coords[:]
+
+        m0 = Function(name='m0', grid=v.grid, space_order=so)
+        dm1 = Function(name='dm1', grid=v.grid, space_order=so)
+        m0.data[:] = 1.5
+
+        # Model perturbation, box of random values centered on middle of model
+        dm1.data[:] = 0
+        size = 5
+        ns = 2 * size + 1
+        if len(shape) == 2:
+            nx2, nz2 = shape[0]//2, shape[1]//2
+            dm1.data[nx2-size:nx2+size, nz2-size:nz2+size] = \
+                -1 + 2 * np.random.rand(ns, ns)
+        else:
+            nx2, ny2, nz2 = shape[0]//2, shape[1]//2, shape[2]//2
+            nx, ny, nz = shape
+            dm1.data[nx2-size:nx2+size, ny2-size:ny2+size, nz2-size:nz2+size] = \
+                -1 + 2 * np.random.rand(ns, ns, ns)
+
+        # Data perturbation
+        rec1 = Receiver(name='rec1', grid=v.grid, time_range=time_axis,
+                        coordinates=rec_coords)
+        nt, nr = rec1.data.shape
+        rec1.data[:] = np.random.rand(nt, nr)
+
+        # Nonlinear modeling
+        rec0, u0, _ = solver.forward(src0, v=m0, save=nt)
+
+        # Linearized modeling
+        rec2, _, _, _ = solver.jacobian_forward(dm1, src0, v=m0, save=nt)
+        dm2, _, _, _ = solver.jacobian_adjoint(rec1, u0, v=m0, save=nt)
+
+        sum_m = np.dot(dm1.data.reshape(-1), dm2.data.reshape(-1))
+        sum_d = np.dot(rec1.data.reshape(-1), rec2.data.reshape(-1))
+        diff = (sum_m - sum_d) / (sum_m + sum_d)
+        print("\nadjoint J %s (so=%d) sum_m, sum_d, diff; %16.10e %+16.10e %+16.10e" %
+              (shape, so, sum_m, sum_d, diff))
+        assert np.isclose(diff, 0., atol=1.e-11)
+
+    @pytest.mark.skip(reason="temporarily skip")
+    @pytest.mark.parametrize('dtype', dtypes)
+    @pytest.mark.parametrize('so', space_orders)
+    def test_derivative_skew_symmetry(self, dtype, so):
+        """
+        We ensure that the first derivatives constructed with calls like
+            f.dx(x0=x+0.5*x.spacing)
+        Are skew (anti) symmetric. See the notebook ssa_01_iso_implementation.ipynb
+        for more details.
+        """
+        n = 101
+        d = 1.0
+        shape = (n, )
+        origin = (0., )
+        extent = (d * (n-1), )
+
+        # Initialize Devito grid and Functions for input(f1,g1) and output(f2,g2)
+        grid1d = Grid(shape=shape, extent=extent, origin=origin, dtype=dtype)
```

Comment:
I'd just call this `grid` like any other Grid

Reply:
there is a Grid named grid below, and you can end up with wacky state if you run this cell and then don't run everything up to another cell. This drive me kind of crazy in other notebooks, like the correctness one for example I think you pretty much have to run sequentially. 

in another language I might add { } around the block I want to have only its own scope, but no idea how to do that in python.

Reply:
unless there is an objection, I will leave as is. closing